litestar/middleware/_internal/exceptions/middleware.py

```diff
@@ -0,0 +1,336 @@
+    """Middleware used to wrap an ASGIApp inside a try catch block and handle any exceptions raised.
+
+    This used in multiple layers of Litestar.
+    """
+
+    def __init__(
+        self, app: ASGIApp, debug: bool | None, exception_handlers: ExceptionHandlersMap | None = None
+    ) -> None:
+        """Initialize ``ExceptionHandlerMiddleware``.
+
+        Args:
+            app: The ``next`` ASGI app to call.
+            debug: Whether ``debug`` mode is enabled. Deprecated. Debug mode will be inferred from the request scope
+            exception_handlers: A dictionary mapping status codes and/or exception types to handler functions.
+
+        .. deprecated:: 2.0.0
+            The ``debug`` parameter is deprecated. It will be inferred from the request scope
+
+        .. deprecated:: 2.9.0
+            The ``exception_handlers`` parameter is deprecated. It will be inferred from the application or the
+            route handler.
+        """
+        self.app = app
+        self.exception_handlers = exception_handlers
+        self.debug = debug
+
+        if debug is not None:
+            warn_deprecation(
+                "2.0.0",
+                deprecated_name="debug",
+                kind="parameter",
+                info="Debug mode will be inferred from the request scope",
+                removal_in="3.0.0",
+            )
+
+        if exception_handlers is not None:
+            warn_deprecation(
+                "2.9.0",
+                deprecated_name="exception_handlers",
+                kind="parameter",
+                info="It will be inferred from the application or the route handler",
+                removal_in="3.0.0",
+            )
+
+        self._get_debug = self._get_debug_scope if debug is None else lambda *a: debug
+
+    @staticmethod
+    def _get_debug_scope(scope: Scope) -> bool:
+        return scope["app"].debug
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        """ASGI-callable.
+
+        Args:
+            scope: The ASGI connection scope.
+            receive: The ASGI receive function.
+            send: The ASGI send function.
+
+        Returns:
+            None
+        """
+        try:
+            await self.app(scope, receive, send)
+        except Exception as e:  # noqa: BLE001
+            litestar_app = scope["app"]
+
+            if litestar_app.logging_config and (logger := litestar_app.logger):
+                self.handle_exception_logging(logger=logger, logging_config=litestar_app.logging_config, scope=scope)
+
+            for hook in litestar_app.after_exception:
+                await hook(e, scope)
+
+            if litestar_app.pdb_on_exception:
+                pdb.post_mortem()
+
+            if scope["type"] == ScopeType.HTTP:
+                await self.handle_request_exception(
+                    litestar_app=litestar_app, scope=scope, receive=receive, send=send, exc=e
+                )
+            else:
```

Comment:
What about exceptions in the asgi scope?

Reply:
DYM something like `elif ScopeType == ScopeType.ASGI: ...`?

Actually, news to me that we use `ScopeType.ASGI` internally, as its not part of the spec I hadn't anticipated it.

I _think_ its OK, b/c the only place the IDE shows me that we read that value is when instantiating `ASGIRoute`, where we pass it to the `scope_type` parameter of `BaseRoute` - it doesn't look like we ever overwrite the value of "type" within the scope.

Nonetheless, I'll add an e2e test to ensure that exceptions raised out of a mounted asgi app are handled.

Side note: this code hasn't changed in this PR, only moved.

Reply:
This commit adds that test: https://github.com/litestar-org/litestar/pull/3389/commits/1dddfe05d5051d304a1479c99eec8ad71ab51df8

LMK if that's what you meant.

Reply:
> LMK if that's what you meant.

Yup, exactly!